README.md
# InputField

[![](https://img.shields.io/github/release/ReFreezed/InputField.svg)](https://github.com/ReFreezed/InputField/releases/latest)
[![](https://img.shields.io/github/license/ReFreezed/InputField.svg)](https://github.com/ReFreezed/InputField/blob/master/LICENSE.txt)

**InputField** for [LÖVE](https://love2d.org/) enables simple handling of user text input into your program.
The library is a [single file](https://raw.githubusercontent.com/ReFreezed/InputField/master/InputField.lua) with no external dependencies.
[MIT license](LICENSE.txt).

You can download the [latest release](https://github.com/ReFreezed/InputField/releases/latest) or clone the repository.

- [Features](#features)
- [Basic Usage](#basic-usage)



## Features

- Different field types: *Single-line*, *multi-line* (with or without wrapping), *password* (obscured characters).
- Text cursor and navigation (by keyboard and mouse).
- Text selection.
- Scrolling, both vertical and horizontal.
- Text alignment.
- Shortcuts for common operations, like copying selected text or deleting the next word.
- Undo and redo (history).
- Helper functions for rendering.

The library does not do any rendering itself, but provides helper functions for rendering text, cursors and selections.



## Basic Usage

```lua
local InputField = require("InputField")
local field      = InputField("Initial text.")

local fieldX = 80
local fieldY = 50

love.keyboard.setKeyRepeat(true)

function love.keypressed(key, scancode, isRepeat)
	field:keypressed(key, isRepeat)
end
function love.textinput(text)
	field:textinput(text)
end

function love.mousepressed(mx, my, mbutton, pressCount)
	field:mousepressed(mx-fieldX, my-fieldY, mbutton, pressCount)
end
function love.mousemoved(mx, my)
	field:mousemoved(mx-fieldX, my-fieldY)
end
function love.mousereleased(mx, my, mbutton)
	field:mousereleased(mx-fieldX, my-fieldY, mbutton)
end
function love.wheelmoved(dx, dy)
	field:wheelmoved(dx, dy)
end

function love.draw()
	love.graphics.setColor(0, 0, 1)
	for _, x, y, w, h in field:eachSelection() do
		love.graphics.rectangle("fill", fieldX+x, fieldY+y, w, h)
	end

	love.graphics.setColor(1, 1, 1)
	for _, text, x, y in field:eachVisibleLine() do
		love.graphics.print(text, fieldX+x, fieldY+y)
	end

	local x, y, h = field:getCursorLayout()
	love.graphics.rectangle("fill", fieldX+x, fieldY+y, 1, h)
end
```

See the [library file](https://raw.githubusercontent.com/ReFreezed/InputField/master/InputField.lua) for documentation,
or the [examples folder](https://github.com/ReFreezed/InputField/tree/master/examples) for more elaborate example programs.




CHANGELOG.txt
Changelog
InputField

v3.3.1 (2022-08-18)
- Changed: scroll() now returns how much was actually scrolled.
- Changed: wheelmoved() now only returns true if any scrolling actually happened.
- Fixed "font filtering blocks newlines in multi-line fields".
- Fixed "undo/redo counts as editing text even when nothing happened".

v3.3 (2022-03-15)
- Added: Event method wheelmoved() (which can do horizontal scrolling).
- Added: Triple-click-and-drag to select lines.
- Added: Method setType().
- Added: Methods getInfoAtCoords(), getInfoAtCursor(), getInfoAtCharacter().
- Added: Methods getVisibleLine(), getVisibleLineCount().
- Added: Methods getWheelScrollSpeed(), setWheelScrollSpeed().
- Changed: Better/proper shortcuts in macOS.
- Changed: The cursor's original x position is now preserved when navigating vertically multiple lines.
- Changed: Newlines now count as word boundaries.
- Changed: Navigating up on the first line or down on the last line now does nothing in multi-line fields.
- Changed: The second argument for setMouseScrollSpeed() is now optional.
- Fixed "pressing Home on wrapped lines does nothing". It now navigates to before the soft line wrap.
- Fixed "pressing Escape does not stop the dragging".


v3.2 (2022-03-10)
- Added: Text alignment (left/right/center).
- Added: PageUp and PageDown work in multi-line fields.
- Added: Methods getScrollHandles(), getScrollHandleHorizontal(), getScrollHandleVertical().
- Added: Methods canScroll(), canScrollHorizontally(), canScrollVertically().
- Changed: Keyboard interactions are now disabled while dragging.

v3.1 (2022-01-15)
- Added: Double-click-and-drag to select words.
- Added: Shortcut Shift+Delete to cut text.
- Changed: Pressing Escape while dragging stops the dragging.
- Changed: Better Unicode support for word navigation/selection.
- Fixed "word navigation/selection does not work".

v3.0.1 (2022-01-14)
- Fixed: Freezing issue when calling some methods on fields that never had any text.

v3.0 (2022-01-14)
- First public release!


tests/main.lua
--
-- InputField tests
--

local function assertValue(value, expected)
	if value == expected then  return  end
	error("Expected '"..tostring(expected).."', got '"..tostring(value).."'.", 2)
end

io.stdout:setvbuf("no")
io.stderr:setvbuf("no")
collectgarbage("stop")

local InputField = assert(loadfile(love.filesystem.getSource().."/../InputField.lua"))()

--
-- API tests.
--
do
	local field = InputField("foo")

	assertValue(field:getType(),     "normal")
	assertValue(field:isPassword(),  false)
	assertValue(field:isMultiline(), false)

	-- Text.
	assertValue(field:getText(),        "foo")
	assertValue(field:getVisibleText(), "foo")
	field:setText("b\när")
	assertValue(field:getText(),        "bär") -- We're single-line, so no newlines.
	assertValue(field:getTextLength(),  3)

	-- Dimensions.
	field:setDimensions(200, 150)
	local w, h = field:getDimensions()
	assertValue(w, 200)
	assertValue(h, 150)

	field:setWidth(100)
	field:setHeight(50)
	assertValue(field:getWidth(),  100)
	assertValue(field:getHeight(), 50)

	-- Filter function.
	local function filterFunc(char)
		return char == "b"
	end

	field:setFilter(filterFunc)
	assertValue(field:getFilter(), filterFunc)

	field:setText("")
	field:textinput("abc")
	assertValue(field:getText(), "ac")

	field:setFilter(nil)
	assertValue(field:getFilter(), nil)

	-- Font and filtering.
	local imageDataStr = table.concat{
		"\0\0\0\0", "\255\255\255\255",
		"\0\0\0\0", "\255\255\255\255", "\255\255\255\255",
		"\0\0\0\0",
	}
	local imageData = (love.getVersion() < 11) and love.image.newImageData(#imageDataStr/4, 1, imageDataStr) or love.image.newImageData(#imageDataStr/4, 1, "rgba8", imageDataStr)
	local font      = love.graphics.newImageFont(imageData, "12")

	field:setFontFilteringActive(true)
	assertValue(field:isFontFilteringActive(), true)

	local defaultFont = field:getFont()
	field:setFont(font)
	assertValue(field:getFont(), font)

	field:setText("")
	field:textinput("0123210")
	assertValue(field:getText(), "1221")

	field:setFontFilteringActive(false)
	assertValue(field:isFontFilteringActive(), false)

	field:setFont(defaultFont)

	-- Editable.
	field:setText("")
	field:textinput("a")
	assertValue(field:getText(), "a")

	field:setEditable(false)
	assertValue(field:isEditable(), false)

	field:textinput("b")
	assertValue(field:getText(), "a")

	field:setEditable(true)
	assertValue(field:isEditable(), true)

	field:textinput("c")
	assertValue(field:getText(), "ac")

	-- Cursor.
	field:setText("foo")
	field:setCursor(1)
	assertValue(field:getCursor(), 1)
	local pos1, pos2 = field:getSelection()
	assertValue(pos1, 1)
	assertValue(pos2, 1)
	assertValue(field:getSelectedVisibleText(), "")

	field:moveCursor(1)
	assertValue(field:getCursor(), 2)
	assertValue(field:getSelectedVisibleText(), "")

	field:moveCursor(-100, "end")
	assertValue(field:getCursor(), 0)
	assertValue(field:getSelectedVisibleText(), "fo")
	assertValue(field:getCursorSelectionSide(), "start")
	assertValue(field:getAnchorSelectionSide(), "end")

	-- Selection and mouse events.
	field:setText("foobar")

	field:setSelection(1, 4)
	local pos1, pos2 = field:getSelection()
	assertValue(pos1, 1)
	assertValue(pos2, 4)
	assertValue(field:getSelectedText(),        "oob")
	assertValue(field:getSelectedVisibleText(), "oob")

	field:mousepressed (0,    0, 1, 1)
	field:mousemoved   (1000, 0)
	field:mousereleased(1000, 0, 1)
	assertValue(field:getSelectedVisibleText(), "foobar")

	field:mousepressed (0,    0, 1, 1)
	field:releaseMouse ()
	field:mousemoved   (1000, 0)
	field:mousereleased(1000, 0, 1)
	assertValue(field:getSelectedVisibleText(), "")

	field:setSelection(0, 0)
	field:selectAll()
	local pos1, pos2 = field:getSelection()
	assertValue(pos1, 0)
	assertValue(pos2, 6)

	-- insert, replace
	field:setText("foobar")
	field:setSelection(1, 3)
	field:insert("ööö")
	assertValue(field:getText(), "föööbar")

	-- Double clicking.
	field:setDoubleClickMaxDelay(1.50)
	assertValue(field:getDoubleClickMaxDelay(), 1.50)

	field:setText("foo bar")
	field:mousepressed (0, 0, 1, 1)
	field:mousereleased(0, 0, 1)
	field:mousepressed (0, 0, 1, 2)
	field:mousereleased(0, 0, 1)
	assertValue(field:getSelectedText(), "foo")

	-- Mouse scroll speed.
	field:setMouseScrollSpeed(1, 2)
	local speedX, speedY = field:getMouseScrollSpeed()
	assertValue(speedX, 1)
	assertValue(speedY, 2)

	field:setMouseScrollSpeedX(3)
	field:setMouseScrollSpeedY(4)
	assertValue(field:getMouseScrollSpeedX(), 3)
	assertValue(field:getMouseScrollSpeedY(), 4)

	-- History.
	field:clearHistory()

	field:setMaxHistory(10)
	assertValue(field:getMaxHistory(), 10)

	-- Blinking.
	assertValue(type(field:getBlinkPhase()), "number")
	field:resetBlinking()

	-- Misc.
	field:update(0)
end

do
	local field = InputField("fo\no", "multiwrap")

	assertValue(field:getType(),     "multiwrap")
	assertValue(field:isPassword(),  false)
	assertValue(field:isMultiline(), true)

	-- Text.
	assertValue(field:getText(),        "fo\no")
	assertValue(field:getVisibleText(), "fo\no")

	-- Type change.
	field:setType("password")

	assertValue(field:getType(),     "password")
	assertValue(field:isPassword(),  true)
	assertValue(field:isMultiline(), false)

	-- Text.
	assertValue(field:getText(),        "foo")
	assertValue(field:getVisibleText(), "***")
end

do
	-- Bug: Calling some methods on a field that never had text causes a freeze.
	local field = InputField("", "multiwrap")
	field:selectAll()
end

-- @Incomplete:
-- canScroll, canScrollHorizontally, canScrollVertically
-- eachVisibleLine, eachSelection
-- getAlignment, setAlignment
-- getCursorLayout
-- getInfoAtCoords, getInfoAtCursor, getInfoAtCharacter.
-- getScroll, getScrollX, getScrollY, setScroll, setScrollX, setScrollY, scroll, scrollToCursor
-- getScrollHandles, getScrollHandleHorizontal, getScrollHandleVertical
-- getScrollLimits
-- getTextDimensions, getTextWidth, getTextHeight
-- getTextOffset
-- getVisibleLine, getVisibleLineCount.
-- getWheelScrollSpeed, setWheelScrollSpeed.
-- keypressed
-- wheelmoved

print("Tests completed successfully!")
love.event.quit()


tools/generateUnicodeInfo.lua
--
-- Unicode info generator
--
-- How to generate:
-- > Put UnicodeData.txt in "local/UnicodeData.txt". (https://unicode.org/Public/UNIDATA/UnicodeData.txt)
-- > Run: lua tools/generateUnicodeInfo.lua
-- > Copy printed info
--

local cpSetByCategory = {}
local maxCp           = 128

for line in io.lines"local/UnicodeData.txt" do
	--[[
		Fields:
		 1.  Codepoint
		 2.  Name
		 3.  General_Category
		 4.  Canonical_Combining_Class
		 5.  Bidi_Class
		 6.  Decomposition_Type
		 7.  Decomposition_Mapping
		 8.  Numeric_Type
		 9.  Numeric_Value
		 10. Bidi_Mirrored
		 11. Unicode_1_Name (Obsolete as of 6.2.0)
		 12. ISO_Comment (Obsolete as of 5.2.0; Deprecated and Stabilized as of 6.0.0)
		 13. Simple_Uppercase_Mapping
		 14. Simple_Lowercase_Mapping
		 15. Simple_Titlecase_Mapping
	]]
	local cp, category = line:match"^(%x+);[^;]*;([^;]*)"

	cp       = tonumber(cp, 16)
	category = category:sub(1, 1) -- Generalize categories more (e.g. make 'Zs' into 'Z').

	cpSetByCategory[category]     = cpSetByCategory[category] or {}
	cpSetByCategory[category][cp] = true

	maxCp = math.max(maxCp, cp)
end

local INDENTATION = string.rep("\t", 1)

local function generateInfoForCharacterType(name, categories)
	local cpSetJoined = {}
	local count       = 0

	for _, category in ipairs(categories) do
		local cpSet = cpSetByCategory[category]
		local lastStartCp

		for cp = 128, maxCp do
			if cpSet[cp] then
				cpSetJoined[cp] = true
				count           = count + 1
			end
		end
	end

	local singles = {}
	local ranges  = {}
	local lastStartCp

	for cp = 128, maxCp do
		if cpSetJoined[cp] then
			if not cpSetJoined[cp-1] then
				lastStartCp = cp
			end
			if not cpSetJoined[cp+1] then
				if cp == lastStartCp then  table.insert(singles, cp)
				else                       table.insert(ranges, {from=lastStartCp, to=cp})  end
			end
		end
	end

	table.sort(singles)
	table.sort(ranges, function(a, b)  return a.from < b.from  end)

	io.write(INDENTATION, "-- ", name, " (", count, ", ", #singles, "+", #ranges, ")\n")

	io.write(INDENTATION, "local UNICODE_", name, " = newSet{")
	for i, cp in ipairs(singles) do
		if i > 1       then  io.write(",")                  end
		if i % 30 == 0 then  io.write("\n\t", INDENTATION)  end
		io.write(cp)
	end
	io.write("}\n")

	io.write(INDENTATION, "local ranges = {")
	for i, range in ipairs(ranges) do
		if i > 1      then  io.write(",")                  end
		if i %12 == 0 then  io.write("\n\t", INDENTATION)  end
		io.write(range.from, ",", range.to)
	end
	io.write("}\n")
	io.write(INDENTATION, "for i = 1, #ranges, 2 do  for cp = ranges[i], ranges[i+1] do UNICODE_", name, "[cp] = true end  end\n")

	io.write("\n")
end

generateInfoForCharacterType("PUNCTUATION", {"P"--[[punctuation]],"S"--[[symbol]]})
generateInfoForCharacterType("WHITESPACE",  {"Z"--[[separator]]})


examples/multipleFields/main.lua
--
-- InputField example program: Multiple fields
--
-- This example has an array of "text input" objects, each with its own
-- InputField instance, and a system for what text input has focus.
--
-- We also do some simple handling of CJK text composition.
-- (Set ENABLE_CJK to true.)
--
require"setup"

local InputField = require"InputField"

local LG = love.graphics
local LK = love.keyboard



--
-- Values.
--

local FIELD_PADDING    = 6
local FONT_LINE_HEIGHT = 1.3
local SCROLLBAR_WIDTH  = 5
local BLINK_INTERVAL   = 0.90

local ENABLE_CJK              = false
local COMPOSITION_BOX_PADDING = 3

local theFont = ENABLE_CJK and LG.newFont("unifont-14.0.02.ttf", 16) or LG.newFont(16)

local textInputs = {
	{
		field     = InputField("Foo, bar?! Foobar...", "normal"),
		x         = 100,
		y         = 50,
		width     = 140,
		height    = theFont:getHeight() + 2*FIELD_PADDING,
		alignment = "left",
	},
	{
		field     = InputField("v3rY 53Cr37", "password"),
		x         = 100,
		y         = 120,
		width     = 140,
		height    = theFont:getHeight() + 2*FIELD_PADDING,
		alignment = "center",
	},
	{
		field = InputField("Lorem ipsum dolor sit amet, consectetur adipiscing elit.\nSed do eiusmod tempor incididunt ut labore et dolore magna aliqua. "
			.. "Sagittis eu volutpat odio facilisis.\n\nAccumsan sit amet nulla facilisi morbi tempus.\nViverra maecenas accumsan lacus vel facilisis volutpat est velit egestas.\n"
			.. "Adipiscing elit duis tristique sollicitudin.\nFacilisi morbi tempus iaculis urna id volutpat lacus.\nDiam quam nulla porttitor massa id neque aliquam.",
			"multinowrap"
		),
		x         = 330,
		y         = 50,
		width     = 350,
		height    = 120,
		alignment = "left",
	},
	{
		field = InputField("Luctus accumsan tortor posuere ac ut consequat. Ultrices vitae auctor eu augue.\n"
			.. "Placerat orci nulla pellentesque dignissim enim sit amet venenatis urna.\n\nConsequat nisl vel pretium lectus quam.\n"
			.. "Elit at imperdiet dui accumsan sit amet nulla facilisi morbi. Fames ac turpis egestas integer eget aliquet nibh praesent.",
			"multiwrap"
		),
		x         = 100,
		y         = 220,
		width     = 580,
		height    = 190,
		alignment = "right",
	},
}



--
-- Setup.
--

LK.setKeyRepeat(true)
theFont:setLineHeight(FONT_LINE_HEIGHT)

for _, textInput in ipairs(textInputs) do
	textInput.field:setFont(theFont)
	textInput.field:setDimensions(textInput.width-2*FIELD_PADDING, textInput.height-2*FIELD_PADDING)
	textInput.field:setAlignment(textInput.alignment)
end

local focusedTextInput = textInputs[1] -- Nil means no focus.

local textComposition = "" -- For CJK input.



--
-- LÖVE callbacks.
--



local function indexOf(array, value)
	for i = 1, #array do
		if array[i] == value then  return i  end
	end
	return nil -- Value is not in array.
end

local function isPointInsideRectangle(pointX,pointY, rectX,rectY, rectW,rectH)
	return pointX >= rectX and pointY >= rectY and pointX < rectX+rectW and pointY < rectY+rectH
end

local function getTextInputAtCoords(x, y)
	for textInputNumber, textInput in ipairs(textInputs) do
		if isPointInsideRectangle(x, y, textInput.x, textInput.y, textInput.width, textInput.height) then
			return textInput, textInputNumber
		end
	end
	return nil -- No text input at coords.
end



function love.keypressed(key, scancode, isRepeat)
	local fieldIsBusy = (focusedTextInput ~= nil and focusedTextInput.field:isBusy())

	-- First handle keys that override InputFields' behavior.
	if key == "tab" and not fieldIsBusy then
		-- Cycle focused input.
		local i     = indexOf(textInputs, focusedTextInput)
		local shift = LK.isDown("lshift","rshift")

		if     not i then  i = 1
		elseif shift then  i = (i-2) % #textInputs + 1 -- Backwards.
		else               i =  i    % #textInputs + 1 -- Forwards.
		end

		focusedTextInput = textInputs[i]
		focusedTextInput.field:resetBlinking()

	-- Then handle focused InputField (if there is one).
	elseif focusedTextInput and focusedTextInput.field:keypressed(key, isRepeat) then
		-- Event was handled.

	-- Lastly handle keys for when no InputField has focus or the key wasn't handled by the library.
	elseif key == "escape" and not fieldIsBusy then
		love.event.quit()
	end
end

function love.textinput(text)
	if focusedTextInput then
		focusedTextInput.field:textinput(text)
	end
end

if ENABLE_CJK then
	function love.textedited(text, start, length)
		textComposition = text
	end
end



local isPressing         = false
local pressedTextInput   = nil
local pressedMouseButton = 0

function love.mousepressed(mx, my, mbutton, pressCount)
	if not isPressing then
		local hoveredTextInput = getTextInputAtCoords(mx, my)

		if hoveredTextInput then
			focusedTextInput = hoveredTextInput

			isPressing         = true
			pressedTextInput   = focusedTextInput
			pressedMouseButton = mbutton

			local fieldX = pressedTextInput.x + FIELD_PADDING
			local fieldY = pressedTextInput.y + FIELD_PADDING
			pressedTextInput.field:mousepressed(mx-fieldX, my-fieldY, mbutton, pressCount)

		else
			focusedTextInput = nil
		end
	end
end

function love.mousemoved(mx, my, dx, dy)
	if isPressing then
		local fieldX = pressedTextInput.x + FIELD_PADDING
		local fieldY = pressedTextInput.y + FIELD_PADDING
		pressedTextInput.field:mousemoved(mx-fieldX, my-fieldY)
	end
end

function love.mousereleased(mx, my, mbutton, pressCount)
	if isPressing and mbutton == pressedMouseButton then
		local fieldX = pressedTextInput.x + FIELD_PADDING
		local fieldY = pressedTextInput.y + FIELD_PADDING
		pressedTextInput.field:mousereleased(mx-fieldX, my-fieldY, mbutton)
		isPressing = false
	end
end

function love.wheelmoved(dx, dy)
	-- Scroll field under mouse.
	local hoveredTextInput = getTextInputAtCoords(love.mouse.getPosition())

	if hoveredTextInput then
		hoveredTextInput.field:wheelmoved(dx, dy)
	end
end



function love.update(dt)
	if focusedTextInput then
		focusedTextInput.field:update(dt)
	end
end



local extraFont = LG.newFont(12)

function love.draw()
	local drawStartTime = love.timer.getTime()
	LG.clear(.25, .25, .25, 1)

	for i, textInput in ipairs(textInputs) do
		--
		-- Input field.
		--
		local field    = textInput.field
		local fieldX   = textInput.x + FIELD_PADDING
		local fieldY   = textInput.y + FIELD_PADDING
		local hasFocus = (textInput == focusedTextInput)

		-- Field info.
		local text = i .. ", " .. field:getType() .. ", align=" .. field:getAlignment()
		local y    = textInput.y - 3 - extraFont:getHeight()
		LG.setFont(extraFont)
		LG.setColor(1, 1, 1, .5)
		LG.print(text, textInput.x, y)

		-- Background.
		LG.setColor(0, 0, 0)
		LG.rectangle("fill", textInput.x, textInput.y, textInput.width, textInput.height)

		-- Contents.
		do
			LG.setScissor(textInput.x, textInput.y, textInput.width, textInput.height)

			-- Selection.
			if hasFocus then
				LG.setColor(.2, .2, 1)
			else
				LG.setColor(1, 1, 1, .3)
			end
			for _, x, y, w, h in field:eachSelection() do
				LG.rectangle("fill", fieldX+x, fieldY+y, w, h)
			end

			-- Text.
			LG.setFont(field:getFont())
			LG.setColor(1, 1, 1, (hasFocus and 1 or .8))
			for _, line, x, y in field:eachVisibleLine() do
				LG.print(line, fieldX+x, fieldY+y)
			end

			-- Cursor.
			if hasFocus and (field:getBlinkPhase() / BLINK_INTERVAL) % 1 < .5 then
				local w       = 2
				local x, y, h = field:getCursorLayout()
				LG.setColor(1, 1, 1)
				LG.rectangle("fill", fieldX+x-w/2, fieldY+y, w, h)
			end

			LG.setScissor()
		end

		--
		-- Scrollbars.
		--
		local canScrollH, canScrollV                 = field:canScroll()
		local hOffset, hCoverage, vOffset, vCoverage = field:getScrollHandles()

		local hHandleLength = hCoverage * textInput.width
		local vHandleLength = vCoverage * textInput.height
		local hHandlePos    = hOffset   * textInput.width
		local vHandlePos    = vOffset   * textInput.height

		-- Backgrounds.
		LG.setColor(0, 0, 0, .3)
		if canScrollV then  LG.rectangle("fill", textInput.x+textInput.width, textInput.y,  SCROLLBAR_WIDTH, textInput.height)  end -- Vertical scrollbar.
		if canScrollH then  LG.rectangle("fill", textInput.x, textInput.y+textInput.height, textInput.width, SCROLLBAR_WIDTH )  end -- Horizontal scrollbar.

		-- Handles.
		LG.setColor(.7, .7, .7)
		if canScrollV then  LG.rectangle("fill", textInput.x+textInput.width, textInput.y+vHandlePos,  SCROLLBAR_WIDTH, vHandleLength)  end -- Vertical scrollbar.
		if canScrollH then  LG.rectangle("fill", textInput.x+hHandlePos, textInput.y+textInput.height, hHandleLength, SCROLLBAR_WIDTH)  end -- Horizontal scrollbar.

		--
		-- Focus indication outline.
		--
		if hasFocus then
			local lineWidth = 2

			local x = textInput.x      - lineWidth/2
			local y = textInput.y      - lineWidth/2
			local w = textInput.width  + lineWidth
			local h = textInput.height + lineWidth

			if canScrollV then  w = w + SCROLLBAR_WIDTH  end
			if canScrollH then  h = h + SCROLLBAR_WIDTH  end

			LG.setColor(1, 1, 0, .4)
			LG.setLineWidth(lineWidth)
			LG.rectangle("line", x, y, w, h)
		end
	end

	--
	-- CJK text composition box.
	--
	if ENABLE_CJK and textComposition ~= "" and focusedTextInput then
		local field   = focusedTextInput.field
		local fieldX  = focusedTextInput.x + FIELD_PADDING
		local fieldY  = focusedTextInput.y + FIELD_PADDING
		local w       = field:getFont():getWidth(textComposition)
		local x, y, h = field:getCursorLayout() -- Render the composed text near the cursor.

		x = x - COMPOSITION_BOX_PADDING
		y = y + h
		w = w + 2*COMPOSITION_BOX_PADDING
		h = h + 2*COMPOSITION_BOX_PADDING

		LG.setColor(.2, .2, 1)
		LG.rectangle("fill", fieldX+x, fieldY+y, w, h)
		LG.setColor(1, 1, 1)
		LG.print(textComposition, fieldX+x+COMPOSITION_BOX_PADDING, fieldY+y+COMPOSITION_BOX_PADDING)
	end

	--
	-- Stats and information.
	--
	local drawTime = love.timer.getTime() - drawStartTime

	LG.setFont(extraFont)
	LG.setColor(1, 1, 1, .5)

	local hoveredTextInput, textInputNumber = getTextInputAtCoords(love.mouse.getPosition())

	if hoveredTextInput then
		local field    = hoveredTextInput.field
		local fieldX   = hoveredTextInput.x + FIELD_PADDING
		local fieldY   = hoveredTextInput.y + FIELD_PADDING
		local mx, my   = love.mouse.getPosition()
		local info     = field:getInfoAtCoords(mx-fieldX, my-fieldY)
		local charText = "none"

		if info.hasText then
			local charInfo = field:getInfoAtCharacter(info.characterPosition)
			charText       = string.format("position=%d, character='%s'", info.characterPosition, charInfo.character)
		end

		local text = string.format(
			"Field #%d, type=%s, alignment=%s, busy=%s\n"
			.. "Cursor: position=%d\n"
			.. "Selection: position=%d, length=%d\n"
			.. "At mouse:\n"
			.. "  Cursor: position=%d\n"
			.. "  Character: %s\n"
			.. "  Line: index=%d, position=%d, length=%d",
			textInputNumber, field:getType(), field:getAlignment(), (field:isBusy() and "yes" or "no"),
			field:getCursor(),
			field:getSelection(), select(2, field:getSelection())-field:getSelection(),
			info.cursorPosition,
			charText,
			info.lineIndex, info.linePosition, require"utf8".len(field:getVisibleLine(info.lineIndex))
		)
		LG.print(text, 3, LG.getHeight()-3-10*extraFont:getHeight())
	end

	local text = string.format(
		"Memory: %.2f MB\n"
		.. "Draw time: %.1f ms",
		collectgarbage"count" / 1024,
		drawTime * 1000
	)
	LG.print(text, 3, LG.getHeight()-3-2*extraFont:getHeight())
end


